Fix #1811: take limit_choices_to into account with FK

adrienbrunet opened the pull request:
Closes https://github.com/encode/django-rest-framework/issues/1811

I created a small repo to test this fix and it works as expected. It is based on https://github.com/encode/django-rest-framework/issues/5358 proposal. (I've done nothing other than formatting the code)

Tests are passing but some more tests for this particular bit should be added, hence the WIP in the title.

(If anyone wants to help...)

adrienbrunet commented:
Finally I took some time to add a simple test for this.
I don't know if it's enough or if we should add more. It's not really elegant though.
Advices welcome.

xordoquy commented on rest_framework/relations.py:
```diff
@@ -169,6 +169,12 @@ def get_queryset(self):
+                **self.parent.Meta.model._meta.get_field(
+                    self.source
+                ).get_limit_choices_to()
+            )
```
Maybe we could skip the filter if there's no `limit_choices_to` ?
Not sure how to do that though

adrienbrunet commented:
I moved my tests to `test_relations_pk` file and updated the code in relation to filter only when `limit_choices_to` is defined. I'm all ears if you have any other requests.

tomchristie commented on rest_framework/relations.py:
```diff
@@ -169,6 +169,12 @@ def get_queryset(self):
             # for the field.
             # Eg: 'MyRelationship(queryset=ExampleModel.objects.all())'
             queryset = queryset.all()
+        if hasattr(self.parent, "Meta") and hasattr(self.parent.Meta, "model"):
```
I'm not keen on this way around, as it feels like it'll end up being brittle, and also wouldn't work with plain serializer classes. In other cases, rather than introspect inside a relationship if it belongs to a modelserializer, we *change* what kind of field gets automatically generated for `ModelSerializer`.

So...

I *think* what we want is for `ModelSerializer.get_fields` to be returning a relational field that has the `queryset` argument already filtered.

This calls into `build_field()`, which for this case calls into `build_relational_field()`, which calls `get_relation_kwargs`. So...

I *think* what we want is for this `get_relation_kwargs ` function... https://github.com/encode/django-rest-framework/blob/master/rest_framework/utils/field_mapping.py#L236

To filter the `queryset` keyword argument appropriately, if `get_limit_choices_to()` returns something.

tomchristie commented:
Thanks for this. Looks like it's working towards a resolution.

adrienbrunet commented:
> Thanks for this. Looks like it's working towards a resolution.

What better feeling than fixing a 5-year old issue ? :grin: 

What do you think about the latest changes ?

tomchristie commented on rest_framework/utils/field_mapping.py:
```diff
@@ -266,6 +266,9 @@ def get_relation_kwargs(field_name, relation_info):
             return kwargs
+        limit_choices_to = model_field.get_limit_choices_to()
+        if limit_choices_to:
+            kwargs['queryset'] = kwargs['queryset'].filter(**limit_choices_to)
```
I think this block should probably go above the `if has_through_model:` switch. Since there's a couple of cases where the `queryset` keyword argument gets popped off.

tomchristie replied:
Or else use `if queryset in kwargs and limit_choices_to:`, which would also do the trick.

adrienbrunet replied:
You're right. I did not think it through. I just wanted to avoid extra work putting it after the read_only but forgot that `queryset` can be popped.

Maybe another solution would be to add a check like `'queryset' in kwargs`?

adrienbrunet replied:
I was too slow to post my comment and did not see yours. :/

tomchristie commented:
Nice one! I think there's one last bit to fix here - making sure that the change still works in cases where the `queryset` keyword argument doesn't exist anymore.

adrienbrunet commented:
I'm switching back my solution to the previous one and adding the queryset in kwargs check. Would you like a rebase to avoid all this back and forth commits ?

tomchristie commented:
@adrianmester No need to switch it - looks good as is.


And no to the rebase - I'll just use Squash and merge.

tomchristie commented:
It's the oldest open issue on our list!

https://github.com/encode/django-rest-framework/issues?page=6&q=is%3Aissue+is%3Aopen

adrienbrunet commented:
Allright. That's good for me. It was a pleasure to be part of the team for the last half hour or so :smiley:

tomchristie commented:
Boom 💥

ulgens commented:
What if `limit_choices_to` is a Q object?
